File: 0289.GameofLife.go
func gameOfLife(board [][]int)  {
    if len(board) == 0 || len(board[0]) == 0 {
        return
    }
    m, n := len(board), len(board[0])
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            life := -board[i][j]
            for x := i-1; x <= i+1; x++ {
                for y := j-1; y <= j+1; y++ {
                    if x >= 0 && x < m && y >= 0 && y < n {
                        life += board[x][y] & 1
                    }
                }
            }
            if life == 3 || (board[i][j] == 1 && life == 2) {
                board[i][j] += 2
            }
        }
    }
    for i := range board {
        for j := range board[i] {
            board[i][j] >>= 1
        }
    }
}

File: 0000-0999/0289.GameofLife.py
class Solution(object):
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        // O(mn) + O(1), cool
        m = len(board)
        n = len(board[0])
        for i in xrange(m):
            for j in xrange(n):
                self.judge(board, m, n, i, j)
        for i in xrange(m):
            for j in xrange(n):
                board[i][j] >>= 1

    def judge(self, board, m, n, i, j):
        live = -board[i][j]
        for ii in xrange(max(i-1, 0), min(i+2, m)):
            for jj in xrange(max(j-1, 0), min(j+2, n)):
                live += board[ii][jj] & 1
        if live == 3 or (board[i][j] == 1 and live == 2):
            board[i][j] += 2

